pydantic/validators.py

```diff
@@ -443,9 +443,13 @@ def make_literal_validator(type_: Any) -> Callable[[Any], Any]:
     allowed_choices_set = set(permitted_choices)
 
     def literal_validator(v: Any) -> Any:
-        if v not in allowed_choices_set:
+        # Ensure to return the value in `Literal`, not the passed value
+        # In some cases they can indeed be different (see `test_literal_validator_str_enum`)
+        for x in allowed_choices_set:
+            if v == x:
+                return x
+        else:
             raise errors.WrongConstantError(given=v, permitted=permitted_choices)
-        return v
```

Comment:
I think this bit of code is equivalent, but without the O(n)-ness.

```python
    def literal_validator(v: Any) -> Any:
        # Ensure to return the value in `Literal`, not the passed value
        # In some cases they can indeed be different (see `test_literal_validator_str_enum`)
        intersection = {v} & allowed_choices_set  # order of operands matters
        if not intersection:
            raise errors.WrongConstantError(given=v, permitted=permitted_choices)
        return intersection.pop()
```

Reply:
You're right! Excellent thanks a bunch 👍

Reply:
I suggest adding a test for multi-value `Literal`s because I suspect the order wouldn't be respected for all intersections:

```py
from enum import Enum

class Foo(str, Enum):
    bar = "bar"

assert ({"bar"} & {Foo.bar}).pop() is Foo.bar
assert ({"bar"} & {Foo.bar, "baz"}).pop() is Foo.bar
```

Reply:
You're absolutely right @layday! There is indeed a weird behaviour with intersection.
```python
from pydantic.typing import all_literal_values, Literal


class Bar(str, Enum):
    FIZ = 'fiz'
    FUZ = 'fuz'

type_ = Literal[Bar.FIZ, Bar.FUZ]

permitted_choices = all_literal_values(type_)
allowed_choices_set = set(permitted_choices)
intersection = allowed_choices_set & {'fiz'}
# intersection == {'fiz'} !!!!
```

@yobiscus I put back my old implementation, which works with many values in `Literal`.

Reply:
I suppose you could use a dictionary instead, that will still be O(1)-like:

```py
choices = {v: v for v in values_in_literal}

try:
     return choices[value]
except KeyError:
    raise ValueError('invalid')
```

Reply:
@PrettyWood 

```python
permitted_choices = all_literal_values(type_)
allowed_choices_set = set(permitted_choices)
intersection = allowed_choices_set & {'fiz'}
# intersection == {'fiz'} !!!!
```

This is because the order of the operands matter. If you flip the operands to `intersection = {'fiz'} & allowed_choices_set`, the correct result is returned.

@layday This explains why your example is returning the "wrong" value as well.

Reply:
That's not what I'm seeing:

```py
In [1]: %paste
from enum import Enum

class Foo(str, Enum):
    bar = "bar"

## -- End pasted text --

In [2]: {"bar"} & {Foo.bar}
Out[2]: {<Foo.bar: 'bar'>}

In [3]: {Foo.bar} & {"bar"}
Out[3]: {'bar'}

In [4]: {Foo.bar, "baz"} & {"bar"}
Out[4]: {'bar'}

In [5]: {"bar"} & {Foo.bar, "baz"}
Out[5]: {'bar'}
```

Reply:
I'm guessing CPython is optimised to pick the value from the shortest set, preferring the RHS operand.  This is definitely not something we'd want to rely on.

Reply:
I have the same as @layday ! Very strange... Does it depend on the python version? 🤔 
Anyway switching to dict always works ! Updated the last commit thanks 👍

Reply:
@layday Thanks for the clarification. I see that now! Bummer.

```python
allowed_choices = {v: v for v in permitted_choices}
```

This always O(n) now, isn't it? But it's possible that CPython's dict comprehension is faster than a for-loop. Not sure about that though.

Reply:
We create the `allowed_choices` *before* in the ModelMetaclass so the validation done by `literal_validator` is still in O(1) ;)

Reply:
Right! Well this solution is much more intuitive than the obscure set intersection one. Nice work!